### ISSUE TITLE
Support for multiple HumHub Sites

### ISSUE THREAD

**Issue by luke-** [humhub/app#79]

**Comment by marc-farre** [comment]
Maybe, in a first time, we could save the previously URLs entered in the form, and show them as a suggestion to avoid having to type the hole URL each time we switch instance?

And if we can go back to the instance URL form without login out, we could switch instance quite easily.

Of course, it's not perfect, mainly for notifications, but if it's simple to develop it can help the time beeing.

**Comment by PrimozRatej** [comment]
Hey @marc-farre, as you suggested, I've prepared a quick demo.

Users can switch between instances using a dropdown list in the TextFormField. When they are connected to an instance, a swipe gesture on the WebView will reveal a back button for quick navigation back to the opener. Reconnecting to logged-in instances is instantaneous (current session inside WebView). Push notifications are configured per instance, so there's no problem here, and the redirect should continue to work. My concern is only how we can indicate to users that this exists, considering the hidden nature of the back button.

https://github.com/humhub/app/assets/10835179/62315a51-85dd-4beb-b02c-7603842f431d


@luke- any thoughts?

**Comment by marc-farre** [comment]
@PrimozRatej Thanks! That's awesome! Maybe it will solve https://github.com/humhub/app/issues/139 ?

Perhaps we could add an entry in the account menu, next to the "Logout" entry, called something like "Return to form URL"?
Of curse, this entry will only be shown in the mobile app (we should have a way in PHP to know if Humhub is currently executed in a browser or the mobile app).

Another idea: adding a "Clear" button on the URL form to quickly remove the current URL and access to the URLs dropdown list?

**Comment by luke-** [comment]
@marc-farre I like the idea of offering an additional menu item "Back to app" or similar when using the app. We can recognize and display this.

But then we have to somehow collect currently logged in HumHub instances and offer a logout (clear) option.

**Comment by PrimozRatej** [comment]
Hey @luke - we could use the JS channels for this and a custom app header, as you mentioned, and the logic will stay on the web side, I like the idea here.

Regarding saving currently logged-in instances, I don't think that would be necessary. Instead, we can simply save the instances that the user has ever connected to, not necessarily logged in.

If the issue you see involves push notifications (those will be sent to the user if the user hasn't logged out of a particular instance), when the user clicks on a specific push notification, they will be redirected to that instance, and then they can log out.

We can save the connected instances on the mobile device itself in local storage.

**Comment by luke-** [comment]
@PrimozRatej Do I understand correctly that we only save the instances that have ever been entered? 

But the user is only logged in to one installation at a time and should therefore only receive push notifications from this one.

Later, we can expand the feature to allow multiple "Logged In" instances in parallel, including push notifications.

**Comment by marc-farre** [comment]
> Users can switch between instances using a dropdown list

@PrimozRatej Do you plan to release this new feature? It would be very useful! Thanks!

**Comment by PrimozRatej** [comment]
Hey @marc-farre, @luke-, do you think now is the time to support this feature?

**Comment by marc-farre** [comment]
@PrimozRatej I would say we must focus on:
- iOs app
- custom branded app
- reliability of the application (testing)
@luke- ?

**Comment by luke-** [comment]
@marc-farre @PrimozRatej 

I think that with multi HumHub support, we have a clear argument in favour of apple app approval.

Here is a mockup masterpiece of how I would see this feature.

![image](https://github.com/user-attachments/assets/b3456c0b-5547-462a-ad46-376951fbb3d7)


What you think? Should we focus on this topic?   (After Branded Apps.)

**Comment by marc-farre** [comment]
If it helps for the Apple store and it's not too much work developing it, then yes of course, it would be great to have.
But I don't know if it's a big value compared to the browser navigation in the eyes of Apple (it's similar to multiple tabs).

For the UX experience, I feel like it's not intuitive to switch from an instance to another.
I think we need somewhere a dropdown accessible when browsing HumHub.

**Comment by luke-** [comment]
> If it helps for the Apple store and it's not too much work developing it, then yes of course, it would be great to have. But I don't know if it's a big value compared to the browser navigation in the eyes of Apple (it's similar to multiple tabs).

Perhaps we can find another feature that is quicker and easier to implement.
The multi-instance mode is not that easy to implement. 

> For the UX experience, I feel like it's not intuitive to switch from an instance to another. I think we need somewhere a dropdown accessible when browsing HumHub.

- If multiple instances are added to the app, the "Instance selection" should always be displayed first at startup.
- It should be possible to switch to the "Instance selection" via the account menu. 

![image](https://github.com/user-attachments/assets/c057a336-9fd2-49fd-a69e-e6494f7c032e)

**Comment by luke-** [comment]
@Eladnarlea I would prefer above Log Out

**Comment by marc-farre** [comment]
I share with how https://www.mightynetworks.com allows switching instances on their mobile app.
They call the button "Switch network" which might be more intuitive to general public than "Switch HumHub instance".
They use a sidebar offcanvas (sort of modabl box on the side, see https://getbootstrap.com/docs/5.3/components/offcanvas/):

![image](https://github.com/user-attachments/assets/26b7ef60-0534-44de-91e7-ea6137f80d08)

![image](https://github.com/user-attachments/assets/5d4c5af9-5753-4935-9444-3146d0227013)

![image](https://github.com/user-attachments/assets/ff254aa7-875f-4171-b40b-d578b4885b6e)

**Comment by Eladnarlea** [comment]
@PrimozRatej I have sent you a pm via Element to discuss possibilities of UI elements to use. After your validation I would post it to this issue.

**Comment by Eladnarlea** [comment]
Please find the suggested design solution below.

### The idea:
1. Show up to three `last logins` to avoid cluttering the interface
2. when clicking `add network` you will be guided through the existing flow
3. `Need help` should still be visible on the first front screen


https://github.com/user-attachments/assets/9909b885-08cf-4c05-ad21-aeb148bb0c28

<img width="330" alt="Screenshot 2024-08-14 at 10 05 44 AM" src="https://github.com/user-attachments/assets/6da33a87-8dbe-4abe-96f8-6d7d6359ef4f">

**Comment by PrimozRatej** [comment]
@Eladnarlea, I can access the favicon through https://community.humhub.com/manifest.json. I assume the network's logo is different, so I think we would need to add it to the manifest or make it accessible through some kind of open API.

**Comment by PrimozRatej** [comment]
We would likely need to add the badge number value to that same API as well.

**Comment by luke-** [comment]
@PrimozRatej @Eladnarlea We should only work with information here that is available in ‘manifest.json’. (Square Icon) This is the only thing that is reliably available for every HumHub installation. Unfortunately, we cannot add custom attributes here as the format is fixed.

We have to add the notification count later via the SilentPush notification.

**Comment by Eladnarlea** [comment]
@luke- @PrimozRatej That's fine. So only the icon will be visible instead of the whole logo - right? @PrimozRatej Do you need new mockups for this, or will that not be necessary?

**Comment by PrimozRatej** [comment]
@Eladnarlea, there's no need. If there are any changes, we can address them in the second iteration.

**Comment by marc-farre** [comment]
@PrimozRatej In older versions, once logged in to a HumHub instance, it was possible to display the app homepage form (to enter another HumHub instance) URL, by swiping right from the left edge (or clicking on the Android return button).
On v1.0.110, it displays "Do you want to exit an App?".
So we need to sign out each time we want to switch to another HumHub instance.

**Comment by PrimozRatej** [comment]
@marc-farre, @luke- 

Sub task of: https://github.com/humhub/app/issues/79

This is the current impl. 

https://github.com/user-attachments/assets/af215d4a-0658-43ff-a278-4d40e25931c3
Once the user navigates to a specific instance, there is no way to return to the opener page, where they could then navigate to another instance from their list.

The idea I had here is to implement a JS channel, similar to what we did for `isHideOpener`. When the user logs out, we navigate them back to the opener page. This works, but it requires the user to log out in order to return.

To allow the user to stay logged in while still switching between instances, we would need a 'Switch Instance' button. I was thinking this could be placed in the menu.

![Screenshot 2024-12-06 at 15 09 00](https://github.com/user-attachments/assets/eacb226c-2f7a-4c58-ae0d-0ce4b6117b77)

When this is clicked, the JS channel switchInstance will be triggered, and we can redirect the user to the native part of the opener. This button will only be shown for `isMobile` users.

NOTE: To maintain backward compatibility, we do not want to show this button to users with an older version of the mobile app that does not yet support 'Multiple HumHub Sites.' For this, I can add a custom header `isMultiInstanceSupport` to ensure the button is only displayed when this is set to true.

**Comment by marc-farre** [comment]
I think that's perfect.

The `isMultiInstanceSupport` will be useful for white labelled apps, because for most of them we will set it to `false`.

@luke- do you approve the idea?
Should I implement this new menu item for 1.17-beta.3?
Name: "Switch network"? Other idea?

Do we keep `Last logins:` or should we change for `My networks:`?
![image](https://github.com/user-attachments/assets/a3c8b89a-27db-454a-8d39-2c616ea13a40)

**Comment by PrimozRatej** [comment]
> The `isMultiInstanceSupport` will be useful for white labelled apps, because for most of them we will set it to `false`.

Ah, I forgot about white-labeled apps. I think we should support a "Switch Instance" button on all instances, including white-labeled ones. This is because in case users use both the opener app and white-labeled apps for same instance.

For white-labeled apps, we can add an additional header: `isWhiteLabeled`.
The logic would then be:

`If isWhiteLabeled == true`: Do not show the "Switch Instance" button.
`If isWhiteLabeled == false and isMultiInstanceSupport == true`: Show the "Switch Instance" button.

> Do we keep `Last logins:` or should we change for `My networks:`?
I think it's better, yes.

**Comment by luke-** [comment]
### A thought on `HideOpener` State

Currently, the app has the “hideOpener” state. In this state, when you press “Back” or open a closed app, the opener is no longer displayed.
 
If we now allow multiple instances, does it really make sense to continue having this state?

### "Switch Network" in User Account Menu

If we decide to add this menu item (might also make sense without the “hideOpener” state - see above), I like the idea with the headers to determine when to show this option.

---

@marc-farre I would like to implement the “Switch Network” in the FCM push module first.
For v18, we could extract general mobile app components into a core module `mobile-app`. What do you think?

**Comment by marc-farre** [comment]
@luke- Yes, that sounds good.

We had a discussion with @PrimozRatej about the `isWhiteLabeled` idea and though we don't need it (the `isMultiInstanceSupport` is enough).

For the `hideOpener`, I'll leave @PrimozRatej answer.

**Comment by luke-** [comment]
Ok. By the way, we already have detection for whitelabel apps via https://github.com/humhub/fcm-push/blob/master/helpers/MobileAppHelper.php#L66-L80

**Comment by PrimozRatej** [comment]
I think we should leave the isHideOpener state as it is. This state determines whether the app should navigate using the WebView or the native opener screen. If we remove it, navigation back to the native opener screen when the user logs out would no longer work. Also, if the user was using a particular instance before closing the app, it should reopen to the last used instance when launched again if we don't track the state we can't do this.

> Ok. By the way, we already have detection for whitelabel apps via https://github.com/humhub/fcm-push/blob/master/helpers/MobileAppHelper.php#L66-L80

I completely forgot about this. Maybe we could use the existing `isAppWithCustomFcm` implementation to display the switch button. Then, we can reuse the `hideOpener` JS channel to navigate the user back to the opener when switching or loging out.

What do you think @luke- ?

**Comment by marc-farre** [comment]
I think we should allow White Labelled mobile apps having multiple instances.
Example: we could have an international organization having one HumHub instance per country.
But they might want one mobile app only for the hole organization.
And the users to be able to log in to several of their HumHub instance (e.g., and administrator manages 2 countries).
We could have many other examples of organizations with several HumHub instances.
So I suggest keeping the `isMultiInstanceSupport` for all mobile apps.

**Comment by luke-** [comment]
@PrimozRatej 

> I think we should leave the isHideOpener state as it is. This state determines whether the app should navigate using the WebView or the native opener screen. If we remove it, navigation back to the native opener screen when the user logs out would no longer work. Also, if the user was using a particular instance before closing the app, it should reopen to the last used instance when launched again if we don't track the state we can't do this.
> 

Ok, then let's keep the `isHideOpener` state.  

> > Ok. By the way, we already have detection for whitelabel apps via https://github.com/humhub/fcm-push/blob/master/helpers/MobileAppHelper.php#L66-L80
> 
> I completely forgot about this. Maybe we could use the existing `isAppWithCustomFcm` implementation to display the switch button. Then, we can reuse the `hideOpener` JS channel to navigate the user back to the opener when switching or loging out.
> 
> What do you think @luke- ?

That's right, actually we could already implement the button using the existing JS channel. As soon we detect the `isMultiInstanceSupport` flag.

@marc-farre Can you implement this in FCM Push?

**Comment by luke-** [comment]
@marc-farre 

> I think we should allow White Labelled mobile apps having multiple instances. Example: we could have an international organization having one HumHub instance per country. But they might want one mobile app only for the hole organization. And the users to be able to log in to several of their HumHub instance (e.g., and administrator manages 2 countries). We could have many other examples of organizations with several HumHub instances. So I suggest keeping the `isMultiInstanceSupport` for all mobile apps.

Good point. But this is something for the future. So we should not completly remove it.

**Comment by marc-farre** [comment]
> @marc-farre Can you implement this in FCM Push?

Yes.

**Comment by marc-farre** [comment]
@luke- @PrimozRatej PR https://github.com/humhub/fcm-push/pull/60

![image](https://github.com/user-attachments/assets/15f33a8a-049d-439e-a7b2-195252d75a5a)

I have renamed `MobileAppHelper::registerLogoutScript()` to `registerShowOpenerScript()` because this method is not about login out, but about displaying the app Opener

**Comment by luke-** [comment]
Looks good for me. Probably we'll rename the link in the next time. But for now it's fine.

Let me know when i should merge and release a new verison.

**Comment by marc-farre** [comment]
@luke- Primoz is currently testing it on an instance of mine. We'll let you know when it's OK to merge.

**Comment by marc-farre** [comment]
@Eladnarlea Shouldn't we have the same button hight?
![image](https://github.com/user-attachments/assets/c4d4d186-c9df-443f-97e5-6951d1f02929)

**Comment by marc-farre** [comment]
@luke- PR https://github.com/humhub/fcm-push/pull/60 is now ready and tested.

I had to change quite a few things:
- The Events const are changed and moved to `MobileAppHelper`
- The Mobile App admin page Action is moved from the `MobileAppController` to the `AdminController`
- The `MobileAppController` is now accessible to everyone and has the action to switch instance
- I added the new `MobileAppHelper::openerState()` to know if the Opener page is opened or not (and close it if opened)

Can you review this please? https://github.com/humhub/fcm-push/pull/60#discussion_r1883967828

**Comment by luke-** [comment]
@marc-farre Thanks. Looks good for me, but I'm not tested it. Just looked over the source.

Btw:  I'm also not that happy with my original implementation using the session variables (etc.). But I can't actually think of a better solution here. Maybe we should revise this again at a later date. The code is currently very difficult to debug/understand. (Set session var, new request, check if, ...)

**Comment by PrimozRatej** [comment]
@marc-farre, @luke- 
Test apk: [app-release.apk.zip](https://github.com/user-attachments/files/18140205/app-release.apk.zip)

In this PR, some major changes were introduced regarding the opener page. I am attaching the APK file so that it can be tested internally before we push it to production.

https://github.com/user-attachments/assets/ec3ad237-7df0-447f-aa57-595c844d72a2

**Comment by marc-farre** [comment]
Just tested it successfully.
2 remarks:

1/ after clicking on the "Switch Network" button, it would be better to display the instance buttons instead of the URL form

![Screenshot_20241215_132202_com.humhub.app.jpg](https://github.com/user-attachments/assets/1d67cbf0-0905-4030-bc98-ae39a73d15fa)

2/ after switching second time, the CUZY. APP button disappear

![Screenshot_20241215_132347_com.humhub.app.jpg](https://github.com/user-attachments/assets/d54e421b-8792-4f99-9ca2-9707f09f2831)

**Comment by luke-** [comment]
I've only seen the video and haven't tested it myself, but it looks really good! That will be a really great feature!

Should @mbumpalumpa take a look regarding design?

**Comment by marc-farre** [comment]
Yes, @mbumpalumpa if you can advise if the button heights are fine or if they should be the same on the same row (for the "Add Network" button, and when the button name takes 2 lines)?

![image](https://github.com/user-attachments/assets/cdc42e4f-f64d-4024-819d-5ebb4faffa76)

![image](https://github.com/user-attachments/assets/e83cc02e-90c4-40ea-9539-5e58cf20b0ec)

**Comment by PrimozRatej** [comment]
https://github.com/humhub/app/releases/tag/v1.1.0
